gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -537,8 +539,32 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
    */
   @Override
   public DirectoryStream<Path> newDirectoryStream(Path dir, Filter<? super Path> filter) {
-    // TODO: Implement me.
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(dir);
+    checkNotNull(filter);
+    String prefix = cloudPath.toString();
+    Iterator<Blob> iterator = storage.list(cloudPath.bucket(), Storage.BlobListOption.prefix(prefix)).iterateAll();
+    final List<Path> ret = new ArrayList<>();
+    while (iterator.hasNext()) {
+      Path p = (Path)(cloudPath.getFileSystem().getPath(iterator.next().name()));
```

Comment:
Was `path` already taken? Longer than `p` and shorter than `pathUnderConsideration` should make both of us happy :)